tap_wikipedia/wikipedia_abstracts_stream.py

```diff
@@ -116,16 +206,110 @@ def get_records(self, context: Dict | None) -> Iterable[Dict]:
+                    img_url = self.__get_wikipedia_record_images(
+                        record["abstract_info"]["url"]
+                    )
+                except Exception:
+                    self.__logger.warning(
+                        f'error getting URL of {record["abstract_info"]["title"]}',  # noqa: E501
+                        exc_info=True,
+                    )
+                    continue
+
+                record["abstract_info"]["image"] = img_url
+                yield record
+
+        # adds a list of categories to the Wikipedia record
+        def add_categories_to_records(records) -> Iterable[Dict]:
+            for record in records:
+                try:
+                    categories = self.__get_wikipedia_record_categories(
+                        record["abstract_info"]["url"]
+                    )
+                except Exception:
+                    self.__logger.warning(
+                        f'error getting category of {record["abstract_info"]["title"]}',  # noqa: E501
+                        exc_info=True,
+                    )
+                    continue
+
+                record["categories"] = categories
+                yield record
+
+        # adds a list of external links to the Wikipedia record
+        def add_external_links_to_records(records):
+            for record in records:
+                try:
+                    external_links = (
+                        self.__get_wikipedia_record_external_links(  # noqa: E501
+                            record["abstract_info"]["title"]
+                        )
+                    )
+                except Exception:
+                    self.__logger.warning(
+                        f'error getting external links of {record["abstract_info"]["title"]}',  # noqa: E501
+                        exc_info=True,
+                    )
+                    continue
+
+                record["externallinks"] = external_links
+                yield record
+
+        # removes unwanted information from the title of  Wikipedia records
+        def clean_wikipedia_titles(records) -> Iterable[Dict]:
+            for record in records:
+                # removes "Wikipedia" from the Wikipedia Title
+                if record["abstract_info"]["title"][:10] == "Wikipedia:":
+                    record["abstract_info"]["title"] = record["abstract_info"][
+                        "title"
+                    ][  # noqa: E501
+                        10:
+                    ].strip()
+
+                yield record
+
+        # removes unwanted information from the abstract of a Wikipedia record
+        def clean_wikipedia_abstracts(records) -> Iterable[Dict]:
+            for record in records:
+                yield record
+
+        for specification in subset_specification:
+            # extract featured Wikipedia Article records
+            if specification == "featured":
+                records = get_featured_records(records)
```

Comment:
I like the assignment here. You could probably do this with a `compose` type of pattern i.e., build a list of enhancer callables (`list[Callable[[Iterable[RecordType]], Iterable[RecordType]]`) then apply them in one go.

Reply:
https://blog.logrocket.com/how-to-create-compose-function-typescript/

Reply:
Cool. I think the formula I settled on was more of a `pipe` instead.

Reply:
I was contemplating using this pattern within `__select_wikipedia_config_callables` as well, but it might look a bit overengineered. Please let me know what you think.